### PR TITLE
Revert raylet to worker GRPC communication back to asio

### PR DESCRIPTION
**Pull request by pcmoritz**
## What do these changes do?

This reverts https://github.com/ray-project/ray/pull/5120 for the release, which caused problems with message reordering (https://github.com/ray-project/ray/issues/5411), raylet-worker heartbeats (https://github.com/ray-project/ray/issues/5343 and https://github.com/ray-project/ray/pull/5120), and other problems we tried to address https://github.com/ray-project/ray/pull/5341, https://github.com/ray-project/ray/pull/5296, https://github.com/ray-project/ray/pull/5313

Ideally, we would keep protobuf for communicating between workers and raylets, but it is better to reset to a known working state and bring the changes back incrementally.

It deploys a workaround for the Java tests (https://github.com/ray-project/ray/pull/5450/files#diff-4eba54416056dabdaeb6d540848b3c62R31), which we can hopefully remove later. The reason was described by @raulchen:

> I'm getting the following error on my machine with your PR. I seems that it's because sometimes raylet won't gracefully exit and thus won't clean up the socket file (Java use a fixed socket file for all tests /tmp/ray/sockets/raylet)
```
libc++abi.dylib: terminating with uncaught exception of type boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >: bind: Address already in use
*** Aborted at 1565873121 (unix time) try "date -d @1565873121" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGABRT (@0x7fff5e05f2c6) received by PID 63811 (TID 0x111f255c0) stack trace: ***
    @     0x7fff5e10fb5d _sigtramp
    @     0x7fff5b1b2998 GCC_except_table51
    @     0x7fff5dfc96a6 abort
    @     0x7fff5b1a5641 abort_message
    @     0x7fff5b1a57c7 default_terminate_handler()
    @     0x7fff5c758eeb _objc_terminate()
    @     0x7fff5b1b119e std::__terminate()
    @     0x7fff5b1b0f86 __cxxabiv1::failed_throw()
    @     0x7fff5b1a3f99 __cxa_throw
    @        0x108704e7b boost::throw_exception<>()
    @        0x108704d94 boost::asio::detail::do_throw_error()
    @        0x108704d23 boost::asio::detail::throw_error()
    @        0x108823f57 boost::asio::basic_socket_acceptor<>::basic_socket_acceptor()
    @        0x108800234 boost::asio::basic_socket_acceptor<>::basic_socket_acceptor()
    @        0x1087ffac9 ray::raylet::Raylet::Raylet()
    @        0x108800e0f ray::raylet::Raylet::Raylet()
    @        0x1086f651e main
    @     0x7fff5df243d5 start
```

> And the reason why the raylet won't exit gracefully is because
```
F0815 20:54:13.702641 275129792 node_manager.cc:418]  Check failed: client_id != gcs_client_->client_table().GetLocalClientId() Exiting because this node manager has mistakenly been marked dead by the monitor.
```
> However, this error occurs randomly, not specific to any particular case.  It seems that raylet would sometimes be blocked.

## Linter

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -263,6 +221,7 @@ cc_library(
     copts = COPTS,
     deps = [
         ":common_cc_proto",
+        ":node_manager_fbs", # TODO
```
Comment (edoakes) [inline BUILD.bazel]: Should remove these

File: java/runtime/src/main/java/org/ray/runtime/raylet/RayletClientImpl.java [hunk 2]
```diff
@@ -39,9 +39,9 @@
   private long client = 0;
 
   // TODO(qwang): JobId parameter can be removed once we embed jobId in driverId.
-  public RayletClientImpl(String schedulerSockName, UniqueId workerId,
+  public RayletClientImpl(String schedulerSockName, UniqueId clientId,
```
Comment (edoakes) [inline java/runtime/src/main/java/org/ray/runtime/raylet/RayletClientImpl.java]: I don't think we should revert the `clientId`->`workerId` change unless this causes problems

File: python/ray/__init__.py [hunk 3]
```diff
@@ -5,13 +5,6 @@
 import os
 import sys
 
-# MUST import ray._raylet before pyarrow to initialize some global variables.
-# It seems the library related to memory allocation in pyarrow will destroy the
-# initialization of grpc if we import pyarrow at first.
-# NOTE(JoeyJiang): See https://github.com/ray-project/ray/issues/5219 for more
-# details.
-import ray._raylet
```
Comment (edoakes) [inline python/ray/__init__.py]: We should leave this in

File: src/ray/core_worker/transport/raylet_transport.cc [hunk 4]
```diff
@@ -58,7 +58,7 @@ void CoreWorkerRayletTaskReceiver::HandleAssignTask(
   // as they use different connections, the `TaskDone` message is sent
   // to raylet via the same connection so the order is guaranteed.
   RAY_UNUSED(raylet_client_->TaskDone());
-  // Send rpc reply.
+  // send rpc reply.
```
Comment (edoakes) [inline src/ray/core_worker/transport/raylet_transport.cc]: shouldn't revert

Comment (edoakes) [reply src/ray/core_worker/transport/raylet_transport.cc]: actually you should just remove this comment...

File: src/ray/raylet/main.cc [hunk 5]
```diff
@@ -171,7 +171,6 @@ int main(int argc, char *argv[]) {
       server.reset();
       gcs_client->Disconnect();
       main_service.stop();
-      RAY_LOG(INFO) << "Raylet server received SIGTERM message, shutting down...";
```
Comment (edoakes) [inline src/ray/raylet/main.cc]: shouldn't revert

File: src/ray/raylet/main.cc [hunk 6]
```diff
@@ -147,7 +147,7 @@ int main(int argc, char *argv[]) {
   RAY_LOG(DEBUG) << "Starting object manager with configuration: \n"
                  << "rpc_service_threads_number = "
                  << object_manager_config.rpc_service_threads_number
-                 << ", object_chunk_size = " << object_manager_config.object_chunk_size;
```
Comment (edoakes) [inline src/ray/raylet/main.cc]: shouldn't revert

File: src/ray/raylet/raylet.h [hunk 7]
```diff
@@ -22,7 +22,7 @@ class NodeManager;
 
 class Raylet {
  public:
-  /// Create a raylet server and listen for local clients.
+  /// Create a node manager server and listen for new clients.
```
Comment (edoakes) [inline src/ray/raylet/raylet.h]: don't revert

File: src/ray/raylet/task_dependency_manager.cc [hunk 8]
```diff
@@ -342,7 +342,7 @@ void TaskDependencyManager::AcquireTaskLease(const TaskID &task_id) {
   // Check that we were able to renew the task lease before the previous one
   // expired.
   if (now_ms > it->second.expires_at) {
-    RAY_LOG(WARNING) << "Task " << task_id << " lease to renew has already expired by "
+    RAY_LOG(WARNING) << "Task lease to renew has already expired by "
```
Comment (edoakes) [inline src/ray/raylet/task_dependency_manager.cc]: don't revert

File: src/ray/raylet/worker.h [hunk 9]
```diff
@@ -49,6 +45,8 @@ class Worker {
   const JobID &GetAssignedJobId() const;
   void AssignActorId(const ActorID &actor_id);
   const ActorID &GetActorId() const;
+  /// Return the worker's connection.
```
Comment (edoakes) [inline src/ray/raylet/worker.h]: remove comment

File: src/ray/rpc/grpc_server.h [hunk 10]
```diff
@@ -80,10 +71,8 @@ class GrpcServer {
   const std::string name_;
   /// Port of this server.
   int port_;
-  /// Indicates whether this server has been closed.
+  /// Flag indicates whether this server has closed.
```
Comment (edoakes) [inline src/ray/rpc/grpc_server.h]: don't revert

**Review by edoakes — APPROVED** [review]
LGTM once all tests pass

File: BUILD.bazel [hunk 11]
```diff
@@ -102,7 +102,7 @@ cc_proto_library(
 
 # === Begin of rpc definitions ===
 
-# gRPC common lib.
+# GRPC common lib.
```
Comment (edoakes) [inline BUILD.bazel]: don't revert

File: python/ray/__init__.py (hunk 3 above)
Comment (pcmoritz) [reply python/ray/__init__.py]: Hmmm, I prefer not to. It should be fixed properly, once grpc is brought back into the client. This kind of stuff is just a ticking time bomb.

**Comment by zhijunfu** [comment]
Thanks for doing this.

It's probably very difficult to preserve all the grpc changes for feature-flag, but sounds like it's simple to preserve the worker-to-raylet proto and rpc client/server files, so that it would be easier to integrate grpc back for worker-to-raylet communication once the ordering issue is resolved.

**Comment by edoakes** [comment]
I don't think we should keep those proto definitions/implementations in master if they aren't being used. It will be just as easy to add them later from the old commits as they're standalone files and we should generally prevent dead code as much as possible.

**Review by raulchen — DISMISSED** [review]
I'm afraid that it's not a good idea to simply revert the whole PR. Because:
1) it'd be very hard to test and verify if gRPC can work in the future. We are likely to have to keep both asio and grpc in the code base for long term. 
2) Besides changing the communication lib, #5120 also includes some nice optimizations. e.g., consolidating duplicated message definitions, changing worker lookup to use `worker_id`, fixing the `SWAP` queue hack, etc (see my in-line comments). 

I'm still concerned about doing the revert could waste a lot of time. Fixing the issue should be much easier. For example,
1) Make `FetchOrReconstruct` sync (`FetchOrReconstruct` is only called when the worked is blocked on waiting objects, so making it sync won't actually hurt perf).
2) Make the requests idempotent as @ericl suggested (but I don't fully understand this solution yet. @ericl could you please elaborate? thanks)
3) Use #5453.

Any of these 3 fixes looks easier than reverting.

However, if you insist on reverting, could you please only change the underlying communication lib, and keep other things (e.g., message definitions in protobuf)? Thanks in advance.

**Comment by robertnishihara** [comment]
@pcmoritz Please explain what the issue with the Java test is and what the workaround is. It's not at all clear from https://github.com/ray-project/ray/pull/5450/files#diff-4eba54416056dabdaeb6d540848b3c62R31.

File: java/test/src/main/java/org/ray/api/test/BaseTest.java [hunk 12]
```diff
@@ -24,7 +24,11 @@ public void setUpBase(Method method) {
     // These files need to be deleted after each test case.
     filesToDelete = ImmutableList.of(
         new File(Ray.getRuntimeContext().getRayletSocketName()),
-        new File(Ray.getRuntimeContext().getObjectStoreSocketName())
+        new File(Ray.getRuntimeContext().getObjectStoreSocketName()),
+        // TODO(pcm): This is a workaround for
+        // https://github.com/ray-project/ray/pull/5450
+        // and should be fixed properly.
```
Comment (robertnishihara) [inline java/test/src/main/java/org/ray/api/test/BaseTest.java]: The issue should be described. The link just points to this PR, not a description of the issue.

File: src/ray/raylet/node_manager.cc [hunk 13]
```diff
@@ -1067,57 +1100,63 @@ void NodeManager::HandleWaitRequest(const rpc::WaitRequest &request,
...
+      [this, client_blocked, client, current_task_id](std::vector<ObjectID> found,
+                                                      std::vector<ObjectID> remaining) {
+        // Write the data.
+        flatbuffers::FlatBufferBuilder fbb;
+        flatbuffers::Offset<protocol::WaitReply> wait_reply = protocol::CreateWaitReply(
+            fbb, to_flatbuf(fbb, found), to_flatbuf(fbb, remaining));
+        fbb.Finish(wait_reply);
+
+        auto status =
+            client->WriteMessage(static_cast<int64_t>(protocol::MessageType::WaitReply),
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: I guess it wasn't like this before, but while we're at it, do you want to make this a `WriteMessageAsync`?

**Comment by pcmoritz** [comment]
According to Hao:
```
I'm getting the following error on my machine with your PR. I seems that it's because sometimes raylet won't gracefully exit and thus won't clean up the socket file (Java use a fixed socket file for all tests /tmp/ray/sockets/raylet)
libc++abi.dylib: terminating with uncaught exception of type boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >: bind: Address already in use
*** Aborted at 1565873121 (unix time) try "date -d @1565873121" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGABRT (@0x7fff5e05f2c6) received by PID 63811 (TID 0x111f255c0) stack trace: ***
    @     0x7fff5e10fb5d _sigtramp
    @     0x7fff5b1b2998 GCC_except_table51
    @     0x7fff5dfc96a6 abort
    @     0x7fff5b1a5641 abort_message
    @     0x7fff5b1a57c7 default_terminate_handler()
    @     0x7fff5c758eeb _objc_terminate()
    @     0x7fff5b1b119e std::__terminate()
    @     0x7fff5b1b0f86 __cxxabiv1::failed_throw()
    @     0x7fff5b1a3f99 __cxa_throw
    @        0x108704e7b boost::throw_exception<>()
    @        0x108704d94 boost::asio::detail::do_throw_error()
    @        0x108704d23 boost::asio::detail::throw_error()
    @        0x108823f57 boost::asio::basic_socket_acceptor<>::basic_socket_acceptor()
    @        0x108800234 boost::asio::basic_socket_acceptor<>::basic_socket_acceptor()
    @        0x1087ffac9 ray::raylet::Raylet::Raylet()
    @        0x108800e0f ray::raylet::Raylet::Raylet()
    @        0x1086f651e main
    @     0x7fff5df243d5 start
```

**Comment by pcmoritz** [comment]
I can't unfortunately reproduce it. The workaround is there to make sure the Java tests are not broken. It should be properly fixed after the release.

File: src/ray/raylet/node_manager.cc [hunk 14]
```diff
@@ -1783,44 +1752,40 @@ bool NodeManager::AssignTask(const Task &task) {
...
-      const auto prev_actor_task_id = spec.PreviousActorTaskDummyObjectId();
-      RAY_CHECK(!prev_actor_task_id.IsNil());
-      // Add the new handle and give it a reference to the finished task's
-      // execution dependency.
-      actor_entry->second.AddHandle(new_handle_id, prev_actor_task_id);
+  auto task_id = spec.TaskId();
+  auto finish_assign_task_callback = [this, worker, task_id](Status status) {
+    if (worker->UsePush()) {
+      // NOTE: we cannot directly call `FinishAssignTask` here because
+      // it assumes the task is in SWAP queue, thus we need to delay invoking this
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc]: It is not actually true that the `SWAP` queue is needed for asio. It was just one solution for when communication with the worker is asynchronous, which is true in both asio and grpc.

I do agree that the pattern in #5120 was an improvement over this. We could try and keep that here, by just moving tasks directly to the `RUNNING` queue in `DispatchTasks` and calling the code in `FinishAssignTask` directly here (vs calling it in a callback).

**Comment by pcmoritz** [comment]
I agree these are good changes, I think we can bring them back together with the structure of the gRPC handlers (which will also require to convert the serialization back to protobuf). It's a bit more work, let's do it in a follow up PR.

**Comment by raulchen** [comment]
I just merged #5370. There're some small conflicts that should be easy to fix. That PR added return status for `RegisterClient` to indicate whether worker registration is successful. As asio doesn't have status, you can add a `bool successful` flag in `RegisterClientReply`.

**Pull request merged**